In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/01-data-cleaning/treatments_cut.csv
/kaggle/input/01-data-cleaning/patients.csv
/kaggle/input/01-data-cleaning/treatments.csv
/kaggle/input/01-data-cleaning/adverse_reactions.csv


**In this dataset, attend patient 500. I've read the dataset, and there are some problems with this dataset**

I'm now trying to clean up the dataset. let's go

In [2]:
patients= pd.read_csv('/kaggle/input/01-data-cleaning/patients.csv')
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [3]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [4]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [5]:
# patients[patients['weight'] <= 100]

patients[patients['weight'] == 48.800000]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [6]:
patients[patients['height'] == 27]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state  zip_code        country                          contact  birthdate  \
4    AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      27  26.1

In [7]:
patients[patients['address'].isna()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [8]:
patients[patients.duplicated(subset= ['given_name', 'surname'])]

# patients[patients.isna().any(axis=1)]

patient_id assigned_sex given_name surname          address      city  \
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state  zip_code        country                      contact birthdate  \
229    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
237    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
244    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
251    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
277    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   

     weight  height   bmi  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [9]:
import warnings
warnings.filterwarnings('ignore')

# Write a Summary for your data `Masum`

This dataset comprises information about 500 patients, of whom 350 participated in a clinical trial. None of the patients were previously using NovoDra (a popular injectable insulin) or Auralin (an investigational oral insulin) as their primary source of insulin. All were experiencing increasing HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After 4 weeks, which isn’t enough time to capture all the changes in HbA1c that the switch can attribute to Auralin or Novodra:

- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

Data about patients who feel some adverse effects is also recorded.

In [10]:
treatments= pd.read_csv('/kaggle/input/01-data-cleaning/treatments.csv')
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [11]:
treatments.sample()

given_name surname    auralin novodra  hba1c_start  hba1c_end  hba1c_change
98     idalia   moore  25u - 31u       -         8.34        7.9           NaN

In [12]:
treatments.sort_values('hba1c_change',na_position='first')

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika     jindrová  41u - 48u          -         7.63       7.20   
2     yukitaka     takenaka          -  39u - 36u         7.68       7.25   
8        saber       ménard          -  54u - 54u         8.08       7.70   
9         asia      woźniak  30u - 36u          -         7.76       7.37   
10      joseph          day  29u - 36u          -         7.70       7.19   
..         ...          ...        ...        ...          ...        ...   
49     jackson      addison          -  42u - 42u         7.99       7.51   
17        gina         cain          -  36u - 36u         7.88       7.40   
138    giovana        rocha          -  23u - 21u         7.87       7.38   
32       laura  ehrlichmann          -  43u - 40u         7.95       7.46   
245         wu         sung          -  47u - 48u         7.61       7.12   

     hba1c_change  
0             NaN  
2             NaN  
8             NaN  
9             NaN  
10            NaN  
..            ...  
49           0.98  
17           0.98  
138          0.99  
32           0.99  
245          0.99  

[280 rows x 7 columns]

In [13]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [14]:
treatments[treatments.isna().any(axis= 1)]

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika     jindrová  41u - 48u          -         7.63       7.20   
2     yukitaka     takenaka          -  39u - 36u         7.68       7.25   
8        saber       ménard          -  54u - 54u         8.08       7.70   
9         asia      woźniak  30u - 36u          -         7.76       7.37   
10      joseph          day  29u - 36u          -         7.70       7.19   
..         ...          ...        ...        ...          ...        ...   
269     hiromu     horikawa          -  47u - 46u         7.77       7.28   
271        leo       vieira          -  30u - 33u         7.74       7.36   
273       kate    wilkinson  36u - 39u          -         7.72       7.20   
274       naja      enoksen  43u - 50u          -         7.98       7.59   
276       john  teichelmann          -  49u - 49u         7.90       7.58   

     hba1c_change  
0             NaN  
2             NaN  
8             NaN  
9             NaN  
10            NaN  
..            ...  
269           NaN  
271           NaN  
273           NaN  
274           NaN  
276           NaN  

[109 rows x 7 columns]

In [15]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [16]:
treatments[treatments.duplicated()]

given_name surname    auralin novodra  hba1c_start  hba1c_end  \
136     joseph     day  29u - 36u       -          7.7       7.19   

     hba1c_change  
136           NaN

In [17]:
treatments_cut= pd.read_csv('/kaggle/input/01-data-cleaning/treatments_cut.csv')

treatments_cut

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0       jožka   resanovič  22u - 30u          -         7.56       7.22   
1   inunnguaq    heilmann  57u - 67u          -         7.85       7.45   
2       alwin    svensson  36u - 39u          -         7.78       7.34   
3         thể       lương          -  61u - 64u         7.64       7.22   
4      amanda     ribeiro  36u - 44u          -         7.85       7.47   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0           0.34  
1            NaN  
2            NaN  
3           0.92  
4           0.38  
..           ...  
65          0.34  
66          0.95  
67          0.30  
68           NaN  
69           NaN  

[70 rows x 7 columns]

In [18]:
treatments_cut.shape

(70, 7)

In [19]:
treatments_cut[treatments_cut.duplicated(subset= ['given_name', 'surname'])]

Empty DataFrame
Columns: [given_name, surname, auralin, novodra, hba1c_start, hba1c_end, hba1c_change]
Index: []

In [20]:
treatments_cut.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     42.000000
mean      7.838000   7.443143      0.518810
std       0.423007   0.418706      0.270719
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.340000
50%       7.730000   7.345000      0.370000
75%       7.860000   7.467500      0.907500
max       9.910000   9.460000      0.970000

In [21]:
adverse_reactions= pd.read_csv('/kaggle/input/01-data-cleaning/adverse_reactions.csv')

In [22]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [23]:
adverse_reactions.duplicated().sum()

np.int64(0)

In [24]:
# export data for manual assessment

with pd.ExcelWriter('clinical_trials.xlsx') as writer:
  patients.to_excel(writer,sheet_name='patients')
  treatments.to_excel(writer,sheet_name='treatments')
  treatments_cut.to_excel(writer,sheet_name='treatment_cut')
  adverse_reactions.to_excel(writer,sheet_name='adverse_reactions')

### Issues with the dataset

1. Dirty Data

  Table - `Patients`

    - patient_id = 9 has misspelled name 'Dsvid' instead of David `accuracy`
    - state col sometimes contain full name and sometimes times abbreviation `consistency`
    - zip code col has entries with 4 digit `validity`
    - data missing for 12 patients in address,city,	state,zip_code	,country,contact `completion`
    - incorrect data type assigned to sex, zip code, birthdate `validity`
    - duplicate entries by the name of John Doe `accuracy`
    - one patient has weight = 48 pounds `accuracy`
    - one patient has height = 27 inches `accuracy`

  Table - `Treatments` & `Treatments_cut`

    - given_name and surname col is is all lower case `consistency`
    - remove u from Auralin and Novadra cols `validity`
    - '-' in novadra and Auralin col treated as nan `validity`
    - missing values in hba1c_change col `completion`
    - 1 duplicate entry by the name Joseph day `accuracy`
    - in hba1c_change 9 instead of 4 `accuracy`

  Table - `Adverse_reactions`

    - given_name and surname are all in lower case `consistency`

2. Messy Data

  Table - `Patients`

    - contact col contains both phone and email

  Table - `Treatments` & `Treatments_cut`

    - Auralin and Novadra col should be split into 2 cols start and end dose
    - merge both the tables

  Table - `Adverse_reactions`

    - This table should not exist independently

# Order of severity
Completeness <- Validity <- Accuracy <- Consistency


# Data Cleaning Order


1. Quality -> Completeness
2. Tidiness
3. Quality -> Validity
4. Quality -> Accuracy
5. Quality -> Consistency

**Steps involved in Data cleaning**

* Define
* Code
* Test

  `Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [25]:
#original data copy()

patients_df= patients.copy()
treatments_df= treatments.copy()
treatments_cut_df= treatments_cut.copy()
adverse_reactions_df= adverse_reactions.copy()


In [26]:
patients_df.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [27]:
patients_df[patients['address'].isna()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [28]:
#code
patients_df.fillna('No Data',inplace= True)

In [29]:
#test
# patients_df[patients_df['address'].isna()]
patients_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       503 non-null    object 
 5   city          503 non-null    object 
 6   state         503 non-null    object 
 7   zip_code      503 non-null    object 
 8   country       503 non-null    object 
 9   contact       503 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(2), int64(2), object(10)
memory usage: 55.1+ KB


In [30]:
treatments_df.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [31]:
treatments_df[treatments_df['hba1c_change'].isna()]

given_name      surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika     jindrová  41u - 48u          -         7.63       7.20   
2     yukitaka     takenaka          -  39u - 36u         7.68       7.25   
8        saber       ménard          -  54u - 54u         8.08       7.70   
9         asia      woźniak  30u - 36u          -         7.76       7.37   
10      joseph          day  29u - 36u          -         7.70       7.19   
..         ...          ...        ...        ...          ...        ...   
269     hiromu     horikawa          -  47u - 46u         7.77       7.28   
271        leo       vieira          -  30u - 33u         7.74       7.36   
273       kate    wilkinson  36u - 39u          -         7.72       7.20   
274       naja      enoksen  43u - 50u          -         7.98       7.59   
276       john  teichelmann          -  49u - 49u         7.90       7.58   

     hba1c_change  
0             NaN  
2             NaN  
8             NaN  
9             NaN  
10            NaN  
..            ...  
269           NaN  
271           NaN  
273           NaN  
274           NaN  
276           NaN  

[109 rows x 7 columns]

In [32]:
treatments_cut_df.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [33]:
treatments_cut_df[treatments_cut_df['hba1c_change'].isna()].head()

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
1  inunnguaq  heilmann  57u - 67u          -         7.85       7.45   
2      alwin  svensson  36u - 39u          -         7.78       7.34   
6       sara     miles  30u - 35u          -         7.53       7.12   
7    frydryk   adamski  63u - 74u          -         7.75       7.27   
8       jane   citizen          -  37u - 38u         7.98       7.60   

   hba1c_change  
1           NaN  
2           NaN  
6           NaN  
7           NaN  
8           NaN

In [34]:
#code

treatments_df['hba1c_change']= treatments_df['hba1c_start'] - treatments_df['hba1c_end']
treatments_cut_df['hba1c_change']= treatments_cut_df['hba1c_start'] - treatments_cut_df['hba1c_end']

In [35]:
#test
treatments_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  280 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [36]:
#test
treatments_cut_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  70 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [37]:
patients_df['contact'].head()

0          951-719-9170ZoeWellish@superrito.com
1         PamelaSHill@cuvox.de+1 (217) 569-3204
2              402-363-6804JaeMDebord@gustr.com
3    PhanBaLiem@jourrapide.com+1 (732) 636-8246
4               334-515-7487TimNeudorf@cuvox.de
Name: contact, dtype: object

In [38]:
#code--> using

patients_df['phone'] = (patients_df['contact'].astype(str)
    .str.extract(r'(\+?\d[\d\s\-\(\)]{8,}\d)')[0]
    .str.replace(r'\D', '', regex=True)
)


patients_df['email'] = (patients_df['contact'].astype(str)
    .str.replace(r'\+?\d[\d\s\-\(\)]{8,}\d', '', regex=True)
    .str.extract(r'([A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,})')[0]
)

In [39]:
patients_df.drop(columns= 'contact', inplace= True)

In [40]:
patients_df.tail(5)

patient_id assigned_sex given_name      surname  \
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                      address           city     state zip_code  \
498       2530 Victoria Court   Milton Mills        ME   3852.0   
499  494 Clarksburg Park Road         Sedona        AZ  86341.0   
500         649 Nutter Street  Overland Park        MO  64110.0   
501  3652 Boone Crockett Lane        Seattle        WA  98109.0   
502         2778 North Avenue           Burr  Nebraska  68324.0   

           country  birthdate  weight  height   bmi        phone  \
498  United States  4/10/1959   181.1      72  24.6   2074770579   
499  United States  3/26/1948   239.6      70  34.4   9282844492   
500  United States  1/13/1971   171.2      67  26.8   8162236007   
501  United States  2/13/1952   176.9      67  27.7  13604432060   
502  United States   5/3/1954   138.2      71  19.3   4028484923   

                                 email  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com

In [41]:
treatments_df= pd.concat([treatments_df, treatments_cut_df])

In [42]:
treatments_df

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0    veronika    jindrová  41u - 48u          -         7.63       7.20   
1      elliot  richardson          -  40u - 45u         7.56       7.09   
2    yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3        skye  gormanston  33u - 36u          -         7.97       7.62   
4      alissa      montez          -  33u - 29u         7.78       7.46   
..        ...         ...        ...        ...          ...        ...   
65     rovzan     kishiev  32u - 37u          -         7.75       7.41   
66      jakob    jakobsen          -  28u - 26u         7.96       7.51   
67      bernd   schneider  48u - 56u          -         7.74       7.44   
68      berta  napolitani          -  42u - 44u         7.68       7.21   
69     armina       sauvé  36u - 46u          -         7.86       7.40   

    hba1c_change  
0           0.43  
1           0.47  
2           0.43  
3           0.35  
4           0.32  
..           ...  
65          0.34  
66          0.45  
67          0.30  
68          0.47  
69          0.46  

[350 rows x 7 columns]

In [43]:
treatments_df= treatments_df.melt(id_vars= ['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], var_name= 'type', value_name= 'dosage_range')

In [44]:
treatments_df = treatments_df[treatments_df['dosage_range'] != '-']

In [45]:
treatments_df['dosage_start'] = treatments_df['dosage_range'].str.split('-').str.get(0)
treatments_df['dosage_end'] = treatments_df['dosage_range'].str.split('-').str.get(1)

In [46]:
treatments_df.drop(columns= 'dosage_range', inplace= True)

In [47]:
treatments_df['dosage_start'] = treatments_df['dosage_start'].str.replace('u', '')
treatments_df['dosage_end'] = treatments_df['dosage_end'].str.replace('u', '')

In [48]:
treatments_df.tail(10)

given_name     surname  hba1c_start  hba1c_end  hba1c_change     type  \
679      satsita   batukayev         7.63       7.25          0.38  novodra   
680      timothy      cotton         7.92       7.52          0.40  novodra   
681      bjørnar      nilsen         7.99       7.70          0.29  novodra   
684        borna   šlezinger         7.55       7.18          0.37  novodra   
685         mary       adams         7.65       7.26          0.39  novodra   
688  christopher    woodward         7.51       7.06          0.45  novodra   
690        maret    sultygov         7.67       7.30          0.37  novodra   
694        lixue       hsueh         9.21       8.80          0.41  novodra   
696        jakob    jakobsen         7.96       7.51          0.45  novodra   
698        berta  napolitani         7.68       7.21          0.47  novodra   

    dosage_start dosage_end  
679          42          42  
680          26          25  
681          36          33  
684          42          41  
685          32          33  
688          55          51  
690          26          23  
694          22          23  
696          28          26  
698          42          44

In [49]:
treatments_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 350 entries, 0 to 698
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   type          350 non-null    object 
 6   dosage_start  350 non-null    object 
 7   dosage_end    350 non-null    object 
dtypes: float64(3), object(5)
memory usage: 24.6+ KB


In [50]:
treatments_df['dosage_start']= treatments_df['dosage_start'].astype(int)
treatments_df['dosage_end']= treatments_df['dosage_end'].astype(int)

In [51]:
treatments_df = treatments_df.merge(adverse_reactions_df, how ='left', on=['given_name','surname'])

In [52]:
treatments_df.sample(350)

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
171         žarka            rap         7.54       7.15          0.39   
200           zak          kelly         7.66       7.26          0.40   
11    brancaleone          russo         8.61       8.18          0.43   
271         valur      bjarkason         9.71       9.41          0.30   
163        regolo          nucci         7.53       7.02          0.51   
..            ...            ...          ...        ...           ...   
290        pamela           hill         9.53       9.10          0.43   
283        þórunn  tryggvadóttir         7.98       7.51          0.47   
248  kwemtochukwu      ogochukwu         7.86       7.53          0.33   
18      władysław      wieczorek         7.92       7.47          0.45   
19       kristján        ingason         7.92       7.57          0.35   

        type  dosage_start  dosage_end adverse_reaction  
171  auralin            35          48              NaN  
200  novodra            38          38              NaN  
11   auralin            53          60              NaN  
271  novodra            31          36              NaN  
163  auralin            51          59              NaN  
..       ...           ...         ...              ...  
290  novodra            27          29              NaN  
283  novodra            30          32              NaN  
248  novodra            34          37              NaN  
18   auralin            24          37              NaN  
19   auralin            44          55              NaN  

[350 rows x 9 columns]

In [53]:
# treatments_df['adverse_reaction']= treatments_df['adverse_reaction'].fillna('No Reaction')
treatments_df['adverse_reaction'].fillna('No Reaction', inplace=True)

In [54]:
treatments_df['adverse_reaction']

0                    No Reaction
1                    No Reaction
2                    No Reaction
3                    No Reaction
4                    No Reaction
                 ...            
345                       nausea
346                  No Reaction
347    injection site discomfort
348                 hypoglycemia
349    injection site discomfort
Name: adverse_reaction, Length: 350, dtype: object

In [55]:
treatments_df.to_csv('treatments_cleaned.csv', index=False)

In [56]:
treatments_df[treatments_df.duplicated()]

given_name surname  hba1c_start  hba1c_end  hba1c_change     type  \
62     joseph     day          7.7       7.19          0.51  auralin   

    dosage_start  dosage_end adverse_reaction  
62            29          36     hypoglycemia

In [57]:
treatments_df = treatments_df.drop_duplicates()

In [58]:
treatments_df.shape

(349, 9)

In [59]:
patients_df.head(10)

patient_id assigned_sex given_name     surname                     address  \
0           1       female        Zoe     Wellish        576 Brown Bear Drive   
1           2       female     Pamela        Hill   2370 University Hill Road   
2           3         male        Jae      Debord       1493 Poling Farm Road   
3           4         male       Liêm        Phan         2335 Webster Street   
4           5         male        Tim     Neudorf        1428 Turkey Pen Lane   
5           6         male     Rafael       Costa          1140 Willis Avenue   
6           7       female       Mary       Adams            3145 Sheila Lane   
7           8       female     Xiuxiu       Chang  2687 Black Oak Hollow Road   
8           9         male      Dsvid  Gustafsson          1790 Nutter Street   
9          10       female     Sophie     Cabrera           3303 Anmoore Road   

                city       state zip_code        country   birthdate  weight  \
0  Rancho California  California  92390.0  United States   7/10/1976   121.7   
1          Armstrong    Illinois  61812.0  United States    4/3/1967   118.8   
2               York    Nebraska  68467.0  United States   2/19/1980   177.8   
3         Woodbridge          NJ   7095.0  United States   7/26/1951   220.9   
4             Dothan          AL  36303.0  United States   2/18/1928   192.3   
5      Daytona Beach     Florida  32114.0  United States   8/31/1931   183.9   
6            Burbank          NV  84728.0  United States  11/19/1969   146.3   
7        Morgan Hill          CA  95037.0  United States   8/13/1958   158.0   
8        Kansas City          MO  64105.0  United States    3/6/1937   163.9   
9           New York    New York  10011.0  United States   12/3/1930   194.7   

   height   bmi        phone                            email  
0      66  19.6   9517199170         ZoeWellish@superrito.com  
1      66  19.2  12175693204             PamelaSHill@cuvox.de  
2      71  24.8   4023636804             JaeMDebord@gustr.com  
3      70  31.7  17326368246        PhanBaLiem@jourrapide.com  
4      27  26.1   3345157487              TimNeudorf@cuvox.de  
5      70  26.4   3863345237     RafaelCardosoCosta@gustr.com  
6      65  24.3   7755335933            MaryBAdams@einrot.com  
7      60  30.9  14087783236           XiuxiuChang@einrot.com  
8      66  26.5   8162659578      DavidGustafsson@armyspy.com  
9      64  33.4  17187959124  SophieCabreraIbarra@teleworm.us

In [60]:
patients_df.iloc[[8]]

patient_id assigned_sex given_name     surname             address  \
8           9         male      Dsvid  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO  64105.0  United States  3/6/1937   163.9      66  26.5   

        phone                        email  
8  8162659578  DavidGustafsson@armyspy.com

In [61]:
patients_df.loc[8, 'given_name']= patients_df.loc[8, 'given_name'].replace('Dsvid', 'David')